streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1826,6 +1845,36 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
    */
   final def zipWithIndex: ZStream[R, E, (A, Int)] =
     mapAccum(0)((index, a) => (index + 1, (a, index)))
+
+  /**
+   * Zips the two streams so that when a value is emitted by either of the two streams,
+   * it is combined with the latest value from the other stream to produce a result.
+   */
+  final def zipWithLatest[R1 <: R, E1 >: E, B, C](that: ZStream[R1, E1, B])(f0: (A, B) => C): ZStream[R1, E1, C] =
+    new ZStream[R1, E1, C] {
+      def process: ZManaged[R1, E1, InputStream[R1, E1, C]] =
+        for {
+          is    <- self.mergeEither(that).process
+          state <- Ref.make[(Option[A], Option[B])]((None, None)).toManaged_
+          pull: InputStream[R1, E1, C] = {
+            def go: InputStream[R1, E1, C] = is.flatMap { i =>
+              state
+                .modify[InputStream[R1, E1, C]] {
+                  case (previousLeft, previousRight) =>
+                    i match {
+                      case Left(a) =>
+                        previousRight.fold(go)(b => InputStream.emit(f0(a, b))) -> (Some(a) -> previousRight)
```

Comment:
Nicely done! Look ma, no queues!

Reply:
my first formulation was using `raceWith` and three times longer 😅 the `merge*` combinators are a godsend!